askomics/libaskomics/source_file/SourceFile.py

```diff
@@ -136,3 +164,170 @@ def is_decimal(self, value):
                 return True
             except ValueError:
                 return False
+
+    def set_forced_column_types(self, types):
+        """
+        Set manually curated types for column
+        """
+
+        self.forced_column_types = types
+
+    def get_preview_turtle():
+        """
+        Get a preview of turtle representation of data in the source file.
+
+        This is a shortcut to get_turtle(preview_only=True)
+        :return: ttl preview
+        """
+        return self.get_turtle(True)
+
+    def get_abstraction(self):
+        # TODO use rdflib
+        """
+        Get the abstraction representing the source file in ttl format
+
+        :return: ttl content for the abstraction
+        """
+
+        ttl = ''
+        ref_entity = self.get_headers()[0]
+
+        # Store the main entity
+        ttl += AbstractedEntity(ref_entity).get_turtle()
+
+        # Store all the relations
+        for key, key_type in self.forced_column_types.items():
+            if key != 0:
+                ttl += AbstractedRelation(key_type, self.get_headers()[key], ref_entity, self.type_dict[key_type]).get_turtle()
+
+        # Store the startpoint status
+        if self.forced_column_types[0] == 'entity_start':
+            ttl += ":" + ref_entity + ' displaySetting:startPoint "true"^^xsd:boolean .\n'
+        else:
+            ttl += ":" + ref_entity + ' displaySetting:attribute "true"^^xsd:boolean .\n'
+
+    def get_domain_knowledge(self):
+        # TODO use rdflib
+        """
+        Get the domain knowledge representing the source file in ttl format
+
+        :return: ttl content for the domain knowledge
+        """
+
+        if self.category_values == None:
+            # FIXME throw an error: we need to call get_turtle first!!
+            log.error('You must call get_turtle before calling get_abstraction')
+            return
+
+        ttl = ''
+
+        for header, categories in self.category_values.items():
+            indent = len(header) * " " + len("displaySetting:has_category") * " " + 3 * " "
+            ttl = ":" + header + " displaySetting:has_category :"
+            ttl += (" , \n" + indent + ":").join(categories) + " .\n"
+
+            for item in categories:
+                ttl += ":" + item + " rdf:type :" + header + " ;\n" + len(item) * " " + "  rdfs:label \"" + item + "\" .\n"
+
+        # TODO we should return the category values, but they need to be generated first by get_turtle (we should not parse the file 2 times)
+        # FIXME what do we do if self.category_values is empty or if it contains data from preview only?
+        # FIXME maybe we could allow to get domain only when full turtle has been generated?
+
+    def get_turtle(self, preview_only=False):
+        # TODO make this a generator to avoid loading all data in memory, and allow writing in streaming mode
+        # TODO use rdflib
```

Comment:
I'm still not sure if rdflib is the right choice here. The serializer consume an indexed graph (SPO/POS/OSP) which we don't need. rdflib expects a graph with expanded prefixes and remove them during the serialization. It's more efficient to write the RDF prefixed from the start.

I have written a small RDF serializer, taking as input a SPO indexed graph (or an iterable of (S, PO) pairs, where PO is represented by a dict mapping from predicate to set of objects). It's really minimal but sufficient for the askomics uses. I plan to use it to generate SPARQL queries, but let me know if you think it can be usefull here, I could integrate it.